File: 1485-minimum-cost-to-make-at-least-one-valid-path-in-a-grid/minimum-cost-to-make-at-least-one-valid-path-in-a-grid.java
class Solution {
    int[][] dirn={{0,0},{0,1},{0,-1},{1,0},{-1,0}};
    int m;
    int n;

    boolean isValid(int x,int y){
        if(x>=m || y>=n || x<0 || y<0) return false;
        return true;
    }

    class Triplet{
        int x;
        int y;
        int weight;
        // int way;

        Triplet(int x,int y,int weight){
           
            this.x=x;
            this.y=y;
            this.weight=weight;
        }
    }
    public int minCost(int[][] grid) {

        PriorityQueue<Triplet> pq=new PriorityQueue<>((a,b)->{
            return a.weight-b.weight;

        });

        m=grid.length;
        n=grid[0].length;

        boolean[][] visited=new boolean[m][n];

        int[][] distance=new int[m][n];
        for(int i=0;i<m;i++) Arrays.fill(distance[i],Integer.MAX_VALUE);



        pq.add(new Triplet(0,0,0));
        // distance[0][0]=0;
        // visited[0][0]=true;

        while(pq.size()>0){
            Triplet t=pq.remove();
            
            int x=t.x;
            int y=t.y;
            int weight=t.weight;

            if(!visited[x][y]){
                visited[x][y]=true;
            }
            else continue;



            // System.out.println(weight+" x"+x+" y"+y);

            if(x==m-1 && y==n-1) return weight;

            

            for(int i=1;i<=4;i++){
                int flag=1;

                int x_new=x+dirn[i][0];
                int y_new=y+dirn[i][1];

                if(grid[x][y]==i) flag=0;


                if(isValid(x_new,y_new) ){
                    if(!visited[x_new][y_new]){
                        pq.add(new Triplet(x_new,y_new,weight+flag));

                    }
                     
                    //  distance[x_new][y_new]=weight+flag;
                    //  visited[x_new][y_new]=true;
                }
            }
        }
        
        return -1;


        
    }
}

File: 1485-minimum-cost-to-make-at-least-one-valid-path-in-a-grid/minimum-cost-to-make-at-least-one-valid-path-in-a-grid.cpp
class Solution {
public:
    using state=pair<int,int>;
    
    #define F first
    #define S second
    
    

    int dx[4]={0,0,1,-1};
    int dy[4]={1,-1,0,0};

    int minCost(vector<vector<int>>& grid) {

        int m=grid.size();
        int n=grid[0].size();


        vector<vector<int>>dist;
        dist.resize(m,vector<int>(n,1e9));

        deque<state>dq;
        dq.push_back({0,0});
        dist[0][0]=0;


        while(dq.size()>0){
            state curr=dq.front();

            int x=curr.F;
            int y=curr.S;
            int sign=grid[x][y];

            dq.pop_front();

            auto isValid=[m,n](int x,int y){
                if(x>=m || y>=n || x<0 || y<0) return false;
                return true;
            };

            for(int i=0;i<4;i++){
                int x_new=x+dx[i];
                int y_new=y+dy[i];

                

                if(isValid(x_new,y_new)){
                    int cost=1;
                    if(sign-1==i)cost=0;
                    if(dist[x_new][y_new]>dist[x][y]+cost){
                        dist[x_new][y_new]=cost+dist[x][y];
                        if(cost==0){
                            dq.push_front({x_new,y_new});
                        }
                        else{
                            dq.push_back({x_new,y_new});
                        }
                    }
                    
                }
                
            }
        }

        return dist[m-1][n-1];


        
    }
};